src/postgres.c

```diff
@@ -223,6 +281,11 @@ postgres_producer_free(Producer *p)
         if (m->conninfo_replica)
             PQputCopyEnd(m->conn_replica, NULL);
     }
+
+    pthread_cancel(m->commit_worker);
+    void *res;
+    pthread_join(m->commit_worker, &res);
```

Comment:
Shouldn't we join the commit worker thread before we call PQputCopyEnd in producer_free?
or else this could lead to a race between the commit worker thread and producer_free as the latter calls PQputCopyEnd without acquiring the commit lock, no? (the logic in main.c:produce possibly avoids that but I can not be sure without going helgrind on this).

In addition it looks to me that we could deduplicate here and call `_commit` in producer_free as well?

Reply:
Good point, I think I got this wrong.
If we first cancel the thread and then commit, there should be no race condition.

Reply:
{l,s}gtm, it'd be great if you confirm this assumption under helgrind before deploy.